addons/purchase_mrp/models/purchase_mrp.py

```diff
@@ -1,44 +1,72 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from odoo import fields, models
+from odoo import api, models
 from odoo.tools import float_compare
 
 
 class PurchaseOrderLine(models.Model):
     _inherit = 'purchase.order.line'
 
-    qty_received = fields.Float(compute='_compute_qty_received', string="Received Qty", store=True)
-
     def _compute_qty_received(self):
-        super(PurchaseOrderLine, self)._compute_qty_received()
+        phantom_lines = self.env['purchase.order.line']
         for line in self.filtered(lambda x: x.move_ids and x.product_id.id not in x.move_ids.mapped('product_id').ids):
             bom = self.env['mrp.bom']._bom_find(product=line.product_id, company_id=line.company_id.id)
             if bom and bom.type == 'phantom':
                 line.qty_received = line._get_bom_delivered(bom=bom)
+                phantom_lines |= line
+        super(PurchaseOrderLine, self.filtered(lambda x:x not in phantom_lines))._compute_qty_received()
 
     def _get_bom_delivered(self, bom=False):
         self.ensure_one()
-        precision = self.env['decimal.precision'].precision_get('Product Unit of Measure')
-
         # In the case of a kit, we need to check if all components are received or not.
         # nothing policy. A product can have several BoMs, we don't know which one was used when the
         # receipt was created.
-        bom_delivered = {}
         if bom:
-            bom_delivered[bom.id] = False
+            move_dict = dict(map(lambda x: (x.product_id.id, 0.0), self.move_ids))
             product_uom_qty_bom = self.product_uom._compute_quantity(self.product_qty, bom.product_uom_id) / bom.product_qty
             boms, lines = bom.explode(self.product_id, product_uom_qty_bom)
             for bom_line, data in lines:
-                qty = 0.0
                 for move in self.move_ids.filtered(lambda x: x.state == 'done' and x.product_id == bom_line.product_id):
-                    qty += move.product_uom._compute_quantity(move.product_uom_qty, bom_line.product_uom_id)
-                if float_compare(qty, data['qty'], precision_digits=precision) < 0:
-                    bom_delivered[bom.id] = False
-                    break
-                else:
-                    bom_delivered[bom.id] = True
-        if bom_delivered and any(bom_delivered.values()):
-            return self.product_qty
-        elif bom_delivered:
-            return 0.0
+                    move_dict[move.product_id.id] += move.product_uom._compute_quantity(move.product_uom_qty, bom_line.product_uom_id)
+            min_transfer = min(move_dict[bom_line.product_id.id] / data['qty'] for bom_line, data in lines)
+            return min_transfer * self.product_qty
+
+    @api.multi
+    def _create_stock_moves(self, picking):
```

Comment:
https://github.com/odoo-dev/odoo/commit/6f724fd30b205d418bd9b5c5fa2dbf236cfd044f#diff-f6e9b4b36d7d79d7ec315d8db6c935f9R613

it will never same as we are unlinking orignal move from action_confirm.
so when we cancel we did not found kit product on moves.